odbc/src/api/diag.rs

```diff
@@ -0,0 +1,159 @@
+use crate::{
+    api::{data::i16_len, definitions::DiagType},
+    errors::ODBCError,
+};
+use odbc_sys::Pointer;
+use odbc_sys::{Char, Integer, SmallInt, SqlReturn, WChar};
+use std::ptr::copy_nonoverlapping;
+
+///
+/// set_sql_state writes the given sql state to the [`output_ptr`].
+///
+/// # Safety
+/// This writes to a raw C-pointer
+///
+pub unsafe fn set_sql_state(sql_state: &str, output_ptr: *mut WChar) {
+    if output_ptr.is_null() {
+        return;
+    }
+    let sql_state = &format!("{}\0", sql_state);
+    let state_u16 = sql_state.encode_utf16().collect::<Vec<u16>>();
+    copy_nonoverlapping(state_u16.as_ptr(), output_ptr, 6);
+}
+
+///
+/// get_diag_rec copies the given ODBC error's diagnostic information
+/// into the provided pointers.
+///
+/// # Safety
+/// This writes to multiple raw C-pointers
+///
+pub unsafe fn get_diag_rec(
+    error: &ODBCError,
+    state: *mut WChar,
+    message_text: Pointer,
+    buffer_length: SmallInt,
+    text_length_ptr: *mut SmallInt,
+    native_error_ptr: *mut Integer,
+    wstring: bool,
+) -> SqlReturn {
+    if !native_error_ptr.is_null() {
+        *native_error_ptr = error.get_native_err_code();
+    }
+    set_sql_state(error.get_sql_state(), state);
+    let message = format!("{}", error);
+    match wstring {
+        true => i16_len::set_output_wstring(
```

Comment:
Just an idea here and we should probably tackle this in a separate JIRA.
I see that you now have a bunch of switch calling either set_output_string or set_output_wstring in a couple of places already.
I feel that this is bloating the code a little and I'm thinking that one way to make this less "bulky" would be to carry the flag `wstring` all the way down to set_output_string and only have 1 switch in this function. WDYT?

Reply:
yeah, I totally agree, especially if (or when) we implement more of the diag field types, there will be more switches on this. Maybe we can make the ticket dependent / related to the ticket for implementing the rest of the fields?

Reply:
It might require a little more refactoring all over, but it could be something to tackle with the other ticket if not done before because I also think this could be part of some smarter refactoring for supporting W and non-W function with minimal code duplication. Can you update the description of [SQL-1151](https://jira.mongodb.org/browse/SQL-1151)?